### PR TITLE
Cache data on leader elected and use in in subsequent reads.

### PR DESCRIPTION
**Pull request by aquamatthias**

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/MarathonModule.scala [hunk 1]
```diff
@@ -53,6 +50,12 @@ object ModuleNames {
   final val NAMED_SERVER_SET_PATH = "SERVER_SET_PATH"
   final val NAMED_SERIALIZE_GROUP_UPDATES = "SERIALIZE_GROUP_UPDATES"
   final val NAMED_HTTP_EVENT_STREAM = "HTTP_EVENT_STREAM"
+
+  final val STORE_APP = "AppStore"
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/MarathonModule.scala]: the other constants are prefixed with "NAMED_". It would be nice to adhere to that convention.

File: src/main/scala/mesosphere/marathon/health/HealthCheck.scala [hunk 2]
```diff
@@ -16,24 +16,23 @@ import scala.concurrent.duration._
 @JsonIgnoreProperties(ignoreUnknown = true)
 case class HealthCheck(
 
-  @FieldJsonDeserialize(contentAs = classOf[java.lang.String]) path: Option[String] = HealthCheck.DefaultPath,
+    @FieldJsonDeserialize(contentAs = classOf[java.lang.String]) path: Option[String] = HealthCheck.DefaultPath,
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/health/HealthCheck.scala]: why this change?

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import mesosphere.util.ThreadPoolContext
+import org.slf4j.LoggerFactory
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.Future
+
+/**
+  * The entity store cache caches all current versions from the underlying store.
+  * Idea:
+  * - old versions (versioned entries) are never cached
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: update the description. (Mention the prefetch on startup)

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def fetch(key: String): Future[Option[T]] = {
+    if (isUnversionedEntry(key)) {
+      Future.successful(cache.get(key))
+    }
+    else {
+      val unversioned = key.substring(0, key.indexOf(':'))
+      cache.get(unversioned) match {
+        case Some(t) if key == unversioned + ':' + t.version => Future.successful(Some(t))
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: maybe add a comment here

File: src/test/scala/mesosphere/marathon/state/AppRepositoryTest.scala [hunk 5]
```diff
@@ -130,7 +130,7 @@ class AppRepositoryTest extends MarathonSpec {
     assert(res.forall(identity), "Should succeed")
 
     verify(store).names()
-    verify(store).expunge("app1")
+    verify(store).expunge("app1", null)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/state/AppRepositoryTest.scala]: that's really strange. We probably should fix this later. Mark it?

**Comment by kolloch** [comment]
We should add tests for the cache.

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  override def modify(key: String, onSuccess: (T) => Unit = _ => ())(update: Update): Future[T] = {
+    def onModified(t: T): Unit = {
+      if (isUnversionedEntry(key)) cache.update(key, t)
+      onSuccess(t)
+    }
+    store.modify(key, onModified)(update)
+  }
+
+  override def names(): Future[Seq[String]] = Future.successful(cache.keys.toSeq)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: That's incorrect. The list of names has to include older versions as well.

File: src/main/scala/mesosphere/marathon/state/MarathonStore.scala [hunk 7]
```diff
@@ -38,27 +39,35 @@ class MarathonStore[S <: MarathonState[_, S]](
...
-        bytesWritten.update(updatedEntity.bytes.length)
-        store.update(updatedEntity)
-      case None =>
-        val created = f(() => newState()).toProtoByteArray
-        bytesWritten.update(created.length)
-        store.create(prefix + key, created)
+  def modify(key: String, onSuccess: (S) => Unit = _ => ())(f: Update): Future[S] = {
+    lockManager.executeSequentially(key) {
+      log.debug(s"Modify $prefix$key")
+      val res = store.load(prefix + key).flatMap {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/MarathonStore.scala]: Generally, this is incorrect.

As described in #2299, writes are not immediately visible. So instead of a fetch here, we should actually do a lookup in our cache. But this is probably an issue for another PR.

File: src/main/scala/mesosphere/marathon/health/HealthCheck.scala (hunk 2 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/health/HealthCheck.scala]: we removed the MarathonState extension - so this is just a reformat?

File: src/main/scala/mesosphere/marathon/state/MarathonStore.scala (hunk 7 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/state/MarathonStore.scala]: That one is really complicated. But I agree.

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala (hunk 6 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: Added the versioned entries as well.

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      cache.get(unversioned) match {
+        case Some(Some(t)) if key == unversioned + ':' + t.version => Future.successful(Some(t))
+        case _ => store.fetch(key)
+      }
+    }
+  }
+
+  override def modify(key: String, onSuccess: (T) => Unit = _ => ())(update: Update): Future[T] = {
+    def onModified(t: T): Unit = {
+      if (isUnversionedEntry(key)) cache.update(key, Some(t))
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: you need to add `None` for versioned entries

File: src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      unversionedNames.foldLeft[Future[Unit]](Future.successful(())) { (completed, nextName) =>
+        completed.flatMap { _ => preloadEntry(nextName) }
+      }
+    }
+
+    def handleEntries(names: Seq[String]): Future[Unit] = {
+      //add keys with None for version entries
+      names.filterNot(isUnversionedEntry).foreach(cache.put(_, None))
+      //add key with loaded values
+      preloadEntries(names)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/EntityStoreCache.scala]: would be cool to make it more obvious that this works on the unversioned stuff

**Comment by aquamatthias** [comment]
Added test cases.

**Comment by aquamatthias** [comment]
Create a new PR that is based on master
